runtime/src/utils/buffer/write.rs

```diff
@@ -176,138 +153,125 @@ impl<B: Blob> Blob for Write<B> {
-            let end_offset = buffer_offset + data_len;
+        // Case 2: Read entirely from the buffer.
+        if offset >= buffer.offset {
+            let start = (offset - buffer.offset) as usize;
+            let end = start + buf_len;
+            assert!(end <= buffer.data.len()); // should hold due to previous check
 
-            if end_offset > inner.buffer.len() {
-                return Err(Error::BlobInsufficientLength);
-            }
-
-            buf.put_slice(&inner.buffer[buffer_offset..end_offset]);
+            buf.put_slice(&buffer.data[start..end]);
             return Ok(buf);
         }
 
-        // Case 3: Read spans both blob and buffer
-        let blob_bytes = (buffer_start - offset) as usize;
-        let buffer_bytes = data_len - blob_bytes;
+        // Case 3: Read spans both blob and buffer.
+        let blob_bytes = (buffer.offset - offset) as usize;
+        let buffer_bytes = buf_len - blob_bytes;
 
-        // Read from blob first
+        // Read from blob first.
         let blob_part = vec![0u8; blob_bytes];
-        let blob_part = inner.blob.read_at(blob_part, offset).await?;
+        let blob_part = self.blob.read_at(blob_part, offset).await?;
 
-        // Copy blob data and buffer data to result
+        // Copy blob data and buffer data to result.
         buf.as_mut()[..blob_bytes].copy_from_slice(blob_part.as_ref());
-        buf.as_mut()[blob_bytes..].copy_from_slice(&inner.buffer[..buffer_bytes]);
+        buf.as_mut()[blob_bytes..].copy_from_slice(&buffer.data[..buffer_bytes]);
 
         Ok(buf)
     }
 
     async fn write_at(&self, buf: impl Into<StableBuf> + Send, offset: u64) -> Result<(), Error> {
-        // Acquire a write lock on the inner state
-        let mut inner = self.inner.write().await;
-
-        // Prepare the buf to be written
+        // Prepare `buf` to be written.
         let buf = buf.into();
-        let data = buf.as_ref();
-        let data_len = data.len();
+        let buf_len = buf.len(); // number of bytes to write
 
-        // Current state of the buffer in the blob
-        let buffer_start = inner.position;
-        let buffer_end = buffer_start + inner.buffer.len() as u64;
+        // Ensure the write doesn't overflow.
+        let end_offset = offset
+            .checked_add(buf_len as u64)
+            .ok_or(Error::OffsetOverflow)?;
 
-        // Case 1: Simple append to buffered data (most common case)
-        if offset == buffer_end {
-            return inner.write(buf).await;
-        }
+        // Acquire a write lock on the buffer.
+        let mut buffer = self.buffer.write().await;
+
+        // Case 1: Simple append to buffered data (most common case).
+        if offset == buffer.size() {
+            // Flush buffer if adding this data would exceed capacity.
+            if buffer.data.len() + buf_len > buffer.capacity {
+                let (buf, offset) = buffer.take_buf();
+                self.blob.write_at(buf, offset).await?;
```

Comment:
one nit here is that if the buffer is cleared and its offset advanced by take_buf() even if this write fails. Before the buffer would be cleared but its offset would not change, which would also be problematic if we tried to continue I think.  In any case you probably have bigger problems if you're getting write errors, so probably not worth trying to make them recoverable.

Reply:
This is the same as before. I suppose we could handle it by just handing it back to the buffer 🤔